src/java.base/share/classes/com/sun/crypto/provider/XWing.java

```diff
@@ -0,0 +1,387 @@
+        try {
+            MessageDigest sha3 = MessageDigest.getInstance("SHA3-256");
+            return sha3.digest(input);
+        } catch (NoSuchAlgorithmException e) {
+            throw new UnsupportedOperationException("JVM does not support SHA3-256", e);
+        }
+    }
+
+    private static KeyPair deriveX25519KeyPair(byte[] sk) {
+        assert sk.length == 32;
+
+        var privateKey = XWing.X25519.privateKey(sk);
+
+        // applying x25519 on secret key and base point yields the public key:
+        byte[] pk = X25519.dh(privateKey, X25519.basePoint());
+
+        // turn public key bytes into an object:
+        var publicKey = X25519.publicKey(pk);
+        return new KeyPair(publicKey, privateKey);
+    }
+
+    private static byte[] shake256(byte[] input, int byteLength) {
+        var digest = new SHA3.SHAKE256(byteLength);
+        digest.update(input);
+        return digest.digest();
+    }
+
+    private static void destroyQuietly(Destroyable destroyable) {
+        if (destroyable != null) {
+            try {
+                destroyable.destroy();
+            } catch (DestroyFailedException e) {
+                // Ignore
+            }
+        }
+    }
+
+    /**
+     * Utility class for X25519 operations.
+     */
+    private static class X25519 {
+
+        private X25519() {}
+
+        public static XECPublicKey basePoint() {
+            record Holder() {
+                static final XECPublicKey INSTANCE = publicKey(BigInteger.valueOf(9L)); // base point 9 (see RFC 7748)
+            }
+            return Holder.INSTANCE;
+        }
+
+        public static XECPublicKey publicKey(byte[] littleEndianU) {
+            if (littleEndianU == null || littleEndianU.length != 32) {
+                throw new IllegalArgumentException("Public key must be 32 bytes long");
+            }
+            byte[] bigEndianU = littleEndianU.clone();
+            ArrayUtil.reverse(bigEndianU);
+            return publicKey(new BigInteger(1, bigEndianU));
+        }
+
+        public static XECPublicKey publicKey(BigInteger u) {
+            try {
+                return (XECPublicKey) keyFactory().generatePublic(new XECPublicKeySpec(NamedParameterSpec.X25519, u));
+            } catch (InvalidKeySpecException e) {
+                throw new AssertionError("Key spec created in-place", e);
+            }
+        }
+
+        public static XECPrivateKey privateKey(byte[] sk) {
+            try {
+                return (XECPrivateKey) keyFactory().generatePrivate(new XECPrivateKeySpec(NamedParameterSpec.X25519, sk));
+            } catch (InvalidKeySpecException e) {
+                throw new IllegalArgumentException("Internal implementation passed unsuitable key", e);
+            }
+        }
+
+        public static byte[] dh(PrivateKey privateKey, PublicKey publicKey) {
+            try {
+                var keyAgreement = KeyAgreement.getInstance("X25519");
```

Comment:
@wangweij shall I hardcode the SunEC provider or leave it open for the implementation to find a suitable provider?

Reply:
obsoleted by 6f2cc25